File: VB/ReportsSilverlight_SaveLoad_Example.Web/ReportService.svc.vb
Imports Microsoft.VisualBasic
Imports System.IO
Imports System.Text.RegularExpressions
Imports System.Web
Imports DevExpress.Data.Utils.ServiceModel
Imports DevExpress.XtraReports.UI
Imports ReportLibrary

Namespace ReportsSilverlight_SaveLoad_Example.Web
	<SilverlightFaultBehavior> _
	Public Class ReportService
		Inherits DevExpress.XtraReports.Service.ReportService
		Private Shared Function GetFilePath(ByVal reportName As String) As String
			reportName = Regex.Replace(reportName, "[., ]", "_")
			Return HttpContext.Current.Server.MapPath(String.Format("Reports/{0}.repx", reportName))
		End Function

		Protected Overrides Sub FillDataSources(ByVal report As XtraReport, ByVal reportName As String, ByVal isDesignActive As Boolean)
			If TypeOf report Is ProductsReport Then
				report.DataSource = New Northwind().Products
			ElseIf TypeOf report Is CategoriesReport Then
				report.DataSource = New Northwind().Categories
			End If
		End Sub

		Protected Overrides Sub SaveReportLayout(ByVal reportName As String, ByVal layoutData() As Byte)
			File.WriteAllBytes(GetFilePath(reportName), layoutData)
		End Sub
		Protected Overrides Function LoadReportLayout(ByVal reportName As String) As Byte()
			Dim fileName = GetFilePath(reportName)
			If File.Exists(fileName) Then
				Return File.ReadAllBytes(fileName)
			End If
			Return MyBase.LoadReportLayout(reportName)
		End Function
	End Class
End Namespace

File: CS/ReportsSilverlight_SaveLoad_Example.Web/ReportService.svc.cs
using System.IO;
using System.Text.RegularExpressions;
using System.Web;
using DevExpress.Data.Utils.ServiceModel;
using DevExpress.XtraReports.UI;
using ReportLibrary;

namespace ReportsSilverlight_SaveLoad_Example.Web {
    [SilverlightFaultBehavior]
    public class ReportService : DevExpress.XtraReports.Service.ReportService {
        static string GetFilePath(string reportName) {
            reportName = Regex.Replace(reportName, "[., ]", "_");
            return HttpContext.Current.Server.MapPath(string.Format("Reports/{0}.repx", reportName));
        }

        protected override void FillDataSources(XtraReport report, string reportName, bool isDesignActive) {
            if (report is ProductsReport)
                report.DataSource = new Northwind().Products;
            else if (report is CategoriesReport)
                report.DataSource = new Northwind().Categories;
        }

        protected override void SaveReportLayout(string reportName, byte[] layoutData) {
            File.WriteAllBytes(GetFilePath(reportName), layoutData);
        }
        protected override byte[] LoadReportLayout(string reportName) {
            var fileName = GetFilePath(reportName);
            if (File.Exists(fileName))
                return File.ReadAllBytes(fileName);
            return base.LoadReportLayout(reportName);
        }
    }
}
